sem/gridrunner.py

```diff
@@ -0,0 +1,163 @@
+from .runner import SimulationRunner
+import os
+import re
+import uuid
+import drmaa
+
+
+class GridRunner(SimulationRunner):
+    """
+    A Runner which can perform simulations in parallel on a Sun Grid Engine.
+    """
+    def run_simulations(self, parameter_list, data_folder, verbose=False):
+        """
+        This function runs multiple simulations in parallel.
+        """
+
+        # Open up a session
+        s = drmaa.Session()
+        s.initialize()
+
+        # Create a job template for each parameter combination
+        jobs = {}
+        for parameter in parameter_list:
+            # Initialize result
+            current_result = {}
+            current_result.update(parameter)
+
+            command = " ".join([self.script_executable] + ['--%s=%s' % (param,
+                                                                        value)
+                                                           for param, value in
+                                                           parameter.items()])
+
+            # Run from dedicated temporary folder
+            current_result['id'] = str(uuid.uuid4())
+            temp_dir = os.path.join(data_folder, current_result['id'])
+            if not os.path.exists(temp_dir):
+                os.makedirs(temp_dir)
+
+            jt = s.createJobTemplate()
+            jt.remoteCommand = os.path.dirname(
+                os.path.abspath(__file__)) + '/run_program.sh'
+            jt.args = [command]
+            jt.jobEnvironment = self.environment
+            jt.workingDirectory = temp_dir
+            output_filename = os.path.join(temp_dir, 'stdout')
+            error_filename = os.path.join(temp_dir, 'stderr')
+            jt.outputPath = ':' + output_filename
+            jt.errorPath = ':' + error_filename
+
+            jobid = s.runJob(jt)
+
+            # Save the template in our dictionary
+            jobs[jobid] = {
+                'template': jt,
+                'result': current_result,
+                'output': output_filename,
+                'error': error_filename,
+                }
+
+        # Check for job completion, yield results when they are ready
+        try:
+
+            while True:
+                if len(jobs) == 0:
+                    # Clean up
+                    raise StopIteration
+
+                for curjob in jobs.keys():
+                    if s.jobStatus(curjob) is drmaa.JobState.DONE:
```

Comment:
Can this function handle the scenarios that an simulation in one grid machine is failed or in an infinite loop?

Reply:
Not yet, in fact this is one of the things that I want to do better in this Runner!